openedx/core/djangoapps/waffle_utils/__init__.py

```diff
@@ -0,0 +1,188 @@
+        """
+        Returns the request cache used by WafflePlus classes.
+        """
+        return get_request_cache('WafflePlus')
+
+
+class WaffleSwitchPlus(WafflePlus):
+    """
+    Waffle Switch helper class that provides native support for namespacing
+    waffle switches and caching within a request.
+
+    """
+    def is_enabled(self, switch_name):
+        """
+        Returns and caches whether the given waffle switch is enabled.
+        """
+        namespaced_switch_name = self._namespaced_name(switch_name)
+        value = self._cached_switches.get(namespaced_switch_name)
+        if value is None:
+            value = switch_is_active(namespaced_switch_name)
+            self._cached_switches[namespaced_switch_name] = value
+        return value
+
+    @contextmanager
+    def override(self, switch_name, active=True):
+        """
+        Overrides the active value for the given switch for the duration of this
+        contextmanager.
+        Note: The value is overridden in the request cache AND in the model.
+        """
+        previous_active = self.is_enabled(switch_name)
+        try:
+            self.override_for_request(switch_name, active)
+            with self.override_in_model(switch_name, active):
+                yield
+        finally:
+            self.override_for_request(switch_name, previous_active)
+
+    def override_for_request(self, switch_name, active=True):
+        """
+        Overrides the active value for the given switch for the remainder of
+        this request (as this is not a context manager).
+        Note: The value is overridden in the request cache, not in the model.
+        """
+        namespaced_switch_name = self._namespaced_name(switch_name)
+        self._cached_switches[namespaced_switch_name] = active
+        log.info(u"%sSwitch '%s' set to %s for request.", self.log_prefix, namespaced_switch_name, active)
+
+    @contextmanager
+    def override_in_model(self, switch_name, active=True):
+        """
+        Overrides the active value for the given switch for the duration of this
+        contextmanager.
+        Note: The value is overridden in the model, not the request cache.
+        """
+        namespaced_switch_name = self._namespaced_name(switch_name)
+        with waffle_override_switch(namespaced_switch_name, active):
+            log.info(u"%sSwitch '%s' set to %s in model.", self.log_prefix, namespaced_switch_name, active)
+            yield
+
+    @property
+    def _cached_switches(self):
+        """
+        Returns cached active values of all switches in this namespace.
+        """
+        return self._all_cached_switches.setdefault(self.namespace, {})
+
+    @property
+    def _all_cached_switches(self):
+        """
+        Returns dictionary of all switches in the request cache,
+        keyed by namespace.
+        """
+        return self._get_request_cache().setdefault('switches', {})
+
+
+class WaffleFlagPlus(WafflePlus):
```

Comment:
I don't like the name ``WaffleFlagPlus``. It doesn't capture for me what is the important distinction, which is that it can handle per-course enabling of features. How about ``PerCourseFeatureFlag`` or maybe just ``CourseFeatureFlag``.

Another option is to generalize the code so that it isn't about courses. We could make it work with any opaque key, so allowing features to be enabled per-program, per-course, or any other context. Maybe we could just call it ``FeatureFlag`` and all feature flags can support a context.